numpy/_core/src/multiarray/item_selection.c

```diff
@@ -3137,58 +3252,111 @@ PyArray_ArgSort(PyArrayObject *op, int axis, NPY_SORTKIND flags)
 {
     PyArrayObject *op2;
     PyObject *ret;
+    PyArrayMethodObject *argsort_method = NULL;
+    PyArrayMethod_StridedLoop *strided_loop = NULL;
+    PyArrayMethod_SortParameters sort_params = {.flags = flags};
+    PyArrayMethod_Context context = {0};
+    PyArray_Descr *loop_descrs[2];
+    NpyAuxData *auxdata = NULL;
+    NPY_ARRAYMETHOD_FLAGS *method_flags = NULL;
+
     PyArray_ArgSortFunc **argsort_table = NULL;
     PyArray_ArgSortFunc *argsort = NULL;
 
     // Zero the NPY_HEAPSORT bit, maps NPY_HEAPSORT to NPY_QUICKSORT
     flags &= ~_NPY_SORT_HEAPSORT;
 
     // Look for type specific functions
-    argsort_table = PyDataType_GetArrFuncs(PyArray_DESCR(op))->argsort;
-    switch (flags) {
-        case NPY_SORT_DEFAULT:
-            argsort = argsort_table[NPY_QUICKSORT];
-            break;
-        case NPY_SORT_STABLE:
-            argsort = argsort_table[NPY_STABLESORT];
-            break;
-        default:
-            break;
-    }
+    argsort_method = NPY_DT_SLOTS(NPY_DTYPE(PyArray_DESCR(op)))->argsort_meth;
+    if (argsort_method != NULL) {
+        PyArray_Descr *descr = PyArray_DESCR(op);
+        PyArray_Descr *odescr = PyArray_DescrFromType(NPY_INTP);
+        if (odescr == NULL) {
+            return NULL;
+        }
```

Comment:
Can't actually fail (unless the typenum is invalid), and we usually just use that assumption.